expression-compiler/src/main/scala-3/dotty/tools/dotc/evaluation/InsertExpression.scala

```diff
@@ -0,0 +1,154 @@
+          case tree: PackageDef =>
+            val transformed = super.transform(tree).asInstanceOf[PackageDef]
+            if (expressionInserted)
+              // set to `false` to prevent inserting `Expression` class in other `PackageDef`s
+              expressionInserted = false
+              cpy.PackageDef(transformed)(
+                transformed.pid,
+                transformed.stats ++ expressionClass.map(
+                  _.withSpan(tree.span)
+                )
+              )
+            else transformed
+          case tree: Template if isOnBreakpoint(tree) =>
+            expressionInserted = true
+            val exprBlock = mkExprBlock(expression, Literal(Constant(())))
+            val newTemplate = cpy.Template(tree)(body = tree.body :+ exprBlock)
+            super.transform(newTemplate)
+          case tree @ DefDef(name, paramss, tpt, _) if isOnBreakpoint(tree) =>
+            expressionInserted = true
+            cpy.DefDef(tree)(
+              name,
+              paramss,
+              tpt,
+              mkExprBlock(expression, tree.rhs)
+            )
+          case tree @ ValDef(name, tpt, _) if isOnBreakpoint(tree) =>
+            expressionInserted = true
+            cpy.ValDef(tree)(name, tpt, mkExprBlock(expression, tree.rhs))
+          case tree if isOnBreakpoint(tree) =>
+            expressionInserted = true
+            val expr = mkExprBlock(expression, tree)
+              .asInstanceOf[Block]
+            expr
+          case tree =>
+            super.transform(tree)
+
+  private def parseExpression(expression: String)(using Context): Tree =
+    val expressionSource =
+      s"""object Expression:
+         |  { $expression }
+         |""".stripMargin
+    val parsedExpression = parseSource("<wrapped-expression>", expressionSource)
+    parsedExpression
+      .asInstanceOf[PackageDef]
+      .stats
+      .head
+      .asInstanceOf[ModuleDef]
+      .impl
+      .body
+      .head
+
+  private def parseExpressionClass(
+      expressionClassSource: String
+  )(using Context): Seq[Tree] =
+    val parsedExpressionClass =
+      parseSource("source", expressionClassSource)
+        .asInstanceOf[PackageDef]
+    parsedExpressionClass.stats
+
+  private def parseSource(name: String, source: String)(using Context): Tree =
+    val parser = Parsers.Parser(SourceFile.virtual(name, source))
+    parser.parse()
+
+  private def isOnBreakpoint(tree: untpd.Tree)(using Context): Boolean =
+    val startLine =
+      if tree.span.exists then tree.sourcePos.startLine + 1 else -1
+    startLine == evalCtx.breakpointLine
+
+  private def mkExprBlock(expr: Tree, tree: Tree)(using
+      Context
+  ): Tree =
+    val exprTermName = termName(evalCtx.expressionClassName.toLowerCase)
+    val ident = Ident(exprTermName)
+    val valDef = ValDef(exprTermName, TypeTree(), expr)
+    if tree.isDef then Block(List(valDef, ident, tree), Literal(Constant(())))
+    else Block(List(valDef, ident), tree)
```

Comment:
In which case can the tree be a def?

It seems to me that no test fails if I remove the case `tree.isDef`:

```suggestion
    Block(List(valDef, ident), tree)
```

Reply:
Hey @tdudzik, What about this comment?